src/huggingface_hub/utils/cache_explorer.py

```diff
@@ -0,0 +1,246 @@
+    def size_on_disk_str(self) -> str:
+        return _format_size(self.size_on_disk)
+
+
+@dataclass
+class CachedRevisionInfo:
+    r"""Information about a revision (a snapshot).
+
+    A revision can be either referenced by 1 or more `refs` or be "detached" (no refs).
+
+    /!\ `size_on_disk` is not necesarily the sum of all file sizes because of possible
+        duplicated files. Also only blobs are taken into account, not the (neglectable)
+        size of folders and symlinks.
+    """
+
+    commit_hash: str
+    size_on_disk: int
+    files: List[CachedFileInfo]  # sorted by name
+    refs: Optional[Set[str]] = None
+
+    @property
+    def size_on_disk_str(self) -> str:
+        return _format_size(self.size_on_disk)
+
+    @property
+    def nb_files(self) -> int:
+        return len(self.files)
+
+
+@dataclass
+class CachedRepoInfo:
+    r"""Information about a cached repository (dataset or model).
+
+    /!\ `size_on_disk` is not necesarily the sum of all revisions sizes because of
+        duplicated files. Also only blobs are taken into account, not the (neglectable)
+        size of folders and symlinks.
+    """
+
+    repo_id: str
+    repo_type: Literal["model", "dataset"]
+    repo_path: Path
+    revisions: Set[CachedRevisionInfo]
+    size_on_disk: int
+    nb_files: int
+
+    @property
+    def size_on_disk_str(self) -> str:
+        """Human-readable sizes"""
+        return _format_size(self.size_on_disk)
+
+
+@dataclass
+class HFCacheInfo:
+    r"""Information about the entire cache repository.
+
+    /!\ Here `size_on_disk` is equal to the sum of all repo sizes (only blobs).
+    """
+
+    datasets: List[CachedRepoInfo]
+    models: List[CachedRepoInfo]
+    size_on_disk: int
+
+    @property
+    def size_on_disk_str(self) -> str:
+        return _format_size(self.size_on_disk)
+
+
+class CorruptedCacheException(Exception):
+    """Exception for any unexpected structure in the huggingface cache."""
+
+
+def scan_cache_dir(
+    cache_dir: Optional[Union["str", Path]] = None
+) -> List[CachedRepoInfo]:
+    """Scan the entire cache directory and return information about it."""
+    if cache_dir is None:
+        cache_dir = HUGGINGFACE_HUB_CACHE
+
+    datasets: List[CachedRepoInfo] = []
+    models: List[CachedRepoInfo] = []
+    for repo_path in Path(cache_dir).iterdir():
+        try:
+            if repo_path.is_dir():
+                repo_info = _scan_cached_repo(repo_path)
+                if repo_info.repo_type == "dataset":
+                    datasets.append(repo_info)
+                elif repo_info.repo_type == "model":
+                    models.append(repo_info)
+                else:
+                    raise ValueError()
+        except CorruptedCacheException as e:
+            print(f"Error while scanning {repo_path}: {e}")
+
+    return HFCacheInfo(
+        datasets=sorted(datasets, key=lambda repo: repo.repo_path),
+        models=sorted(models, key=lambda repo: repo.repo_path),
```

Comment:
the reason you're sorting is for determinism of output across operating systems?

Reply:
Hmm yes and also to be 100% sure that the output is in alphabetical order. I'm thinking about a user that wants to see a specific model in a long list, it will be easier if it's somehow sorted. 
(and also it really cost nothing to sort a list of max ~1000 items)

Reply:
(I would expect `Path.glob("*")` to be already sorted but it's not guaranteed)

Reply:
Actually I changed the dataclasses to "frozen" so that I can store them in sets instead of lists. Makes much more sense and no need to sort them in `scan_cache` helper.

Then only when I print the result in the CLI, I sort the output by path but not before.